x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java

```diff
@@ -44,18 +45,22 @@
 
         final Predicate<JWK> keyUsePredicate = j -> ((j.getKeyUse() == null) || (KeyUse.SIGNATURE.equals(j.getKeyUse())));
         final List<JWK> jwksSig = jwks.stream().filter(keyUsePredicate).toList();
-        LOGGER.trace("JWKs [" + jwksSig.size() + "] after KeyUse [SIGNATURE||null] filter.");
+        LOGGER.debug("JWKs [" + jwksSig.size() + "] after KeyUse [SIGNATURE||null] filter.");
 
         final Predicate<JWK> keyOpPredicate = j -> ((j.getKeyOperations() == null) || (j.getKeyOperations().contains(KeyOperation.VERIFY)));
         final List<JWK> jwksVerify = jwksSig.stream().filter(keyOpPredicate).toList();
-        LOGGER.trace("JWKs [" + jwksVerify.size() + " after KeyOperation [VERIFY||null] filter.");
+        LOGGER.debug("JWKs [" + jwksVerify.size() + " after KeyOperation [VERIFY||null] filter.");
 
         final List<JWK> jwksFiltered = jwksVerify.stream().filter(j -> (algs.stream().anyMatch(a -> isMatch(j, a)))).toList();
-        LOGGER.trace("JWKs [" + jwksFiltered.size() + "] after Algorithms [" + String.join(",", algs) + "] filter.");
+        LOGGER.debug("JWKs [" + jwksFiltered.size() + "] after Algorithms [" + String.join(",", algs) + "] filter.");
 
         final List<String> algsFiltered = algs.stream().filter(a -> (jwksFiltered.stream().anyMatch(j -> isMatch(j, a)))).toList();
-        LOGGER.trace("Algorithms [" + String.join(",", algsFiltered) + "] after remaining JWKs [" + jwksFiltered.size() + "] filter.");
+        LOGGER.debug("Algorithms [" + String.join(",", algsFiltered) + "] after remaining JWKs [" + jwksFiltered.size() + "] filter.");
 
+        if ((jwksFiltered.size() < jwks.size()) || (algsFiltered.size() < algs.size())) {
+            LOGGER.warn("JWKs [" + jwks.size() + "] usable [" + jwksFiltered.size() + "]. See debug logs.");
+            LOGGER.warn("Algorithms [" + String.join(",", algs) + "] usable [" + String.join(",", algsFiltered) + "]. See debug logs.");
```

Comment:
There are two stages for filtering JWKs vs algs.
- Start up: Takes a list of algs from realm config (ex: HS256, RS256, ES256, ES512)
- Authenticate: Takes a single alg from JWT header (ex: ES256), allows ignoring realm algs for current request (i.e. ES512 has wrong curve to verify an ES256 signature)

I debated checking for size==0, or reduction in size. I went with reduction check. I tried to debug an issue in static realm config, and I realized if I had this much trouble, customers will end up creating a lot of support issues. That warning is all about avoiding SDHs.

Here is the issue. I created a static realm with an HMAC key in UTF8 string format, as per OIDC spec. I picked 20 chars (160-bits), but is too short for HS256. The key is filtered during startup. If PKC has at least one key, the realm will start. If I submit a JWT signed by that key with HS256, I may have that JWK and HS256 in the realm, and the realm started OK, but the request will never authenticate. Customers will blame the problem on a bug.

To avoid support issues coming in for that situation, realm startup needs to have a clear log if any JWKs as ignored due to filtering. If the customer searches for ERROR or WARN, they won't find any log message about problems loading that JWK. Using log level WARN will make it stand out.

Reply:
That line of reasoning assumes that customers got through same process you did. I don't think that's necessarily true.

Take the case of PKC. The keys come from a JWKSet read from a file. The algorithms come from a realm setting. 
The JWKSet may contain multiple keys. The admin may intentionally configure the realm to support a subset of the algorithms for which those keys are usable. There's nothing wrong with:
```
xpack.security.authc.realms:
  jwt.my_jwt.pkc_jwkset_path: "/path/to/jwkset-with-many-rsa-keys.json"
  jwt.my_jwt.allowed_signature_algorithms: "RSA512"  
  # ^^ Reject JWTs signed with less than 512 bit keys
```

By logging a warning you are telling the admin that what they're done is wrong. But it's not. It _might_ be a misconfiguration, it's probably useful to tell them that some keys from the JWKTSet aren't usable, but it's not something to yell at them about.
I agree with `info`.

> realm startup needs to have a clear log if any JWKs as ignored due to filtering

I agree, but I don't think this log message achieves that. Anything that tells admins they need to enable debug logs in order to understand what's going on, does meet the "clear log" test for me.

Reply:
Changed two log messages to info. Resolved.